### PR TITLE
[Workflow] Add support for workflows that need to store many tokens in the marking

### LINKED ISSUE #53179: [Workflow] Verify race condition
### Symfony version(s) affected

all

### Description

The behavior of transitions depends on the order in which they are declared.

### How to reproduce

```php
    private static function createWorkflowWithSameNameBackTransition(): Definition
    {
        $places = range('a', 'c');

        $transitions = [];
        $transitions[] = new Transition('a_to_bc', 'a', ['b', 'c']);
        $transitions[] = new Transition('back1', 'b', 'a');
        $transitions[] = new Transition('back1', 'c', 'b');
        $transitions[] = new Transition('back2', 'c', 'b');
        $transitions[] = new Transition('back2', 'b', 'a');

        return new Definition($places, $transitions);

        // The graph looks like:
        //       +--------+
        //   +---| back_N |<------+ +-----------------+
        //   |   +--------+       | |                 |
        //   v                    | v                 |
        // +---+    +---------+  +---+  +---+     +--------+
        // | a | -> | a_to_bc |  | b |  | c | --> | back_N |
        // +---+    +---------+  +---+  +---+     +--------+
        //                |        ^      ^
        //                +--------+------+
    }

    public static function backVariants()
    {
        return [['back1'], ['back2']];
    }

    /**
     * @dataProvider backVariants
     */
    public function testApplyWithSameNameBackTransition($transition)
    {
        $subject = new Subject();
        $definition = $this->createWorkflowWithSameNameBackTransition();
        $workflow = new Workflow($definition, new MethodMarkingStore());

        $marking = $workflow->apply($subject, 'a_to_bc');

        $this->assertFalse($marking->has('a'));
        $this->assertTrue($marking->has('b'));
        $this->assertTrue($marking->has('c'));

        $marking = $workflow->apply($subject, $transition);

        $this->assertTrue($marking->has('a'));
        $this->assertTrue($marking->has('b'));
        $this->assertFalse($marking->has('c'));
    }

```

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #53179
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/Workflow/Marking.php [hunk 1]
```diff
@@ -27,23 +27,64 @@ class Marking
...
     }
 
-    public function mark(string $place): void
+    /**
+     * @param int $nbToken
+     */
+    public function mark(string $place /* , int $nbToken = 1 */): void
     {
-        $this->places[$place] = 1;
+        if (1 === \func_num_args()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/Marking.php]: please use `>` and a ternary operator, see maybe how we do it on 6.4 for inspiration

**Review by nicolas-grekas — COMMENTED** [review]
Can you please label the title of the PR as a new feature, since it is?

File: src/Symfony/Component/Workflow/Marking.php (hunk 1 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/Marking.php]: mwai :| it's really a matter of personal taste :D . Anyway, updated

**Comment by lyrixx** [comment]
PR updated, thanks

File: src/Symfony/Component/Workflow/Marking.php [hunk 2]
```diff
@@ -27,23 +27,56 @@ class Marking
...
+        }
+
+        if (!$this->has($place)) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" is not marked.', $place));
+        }
+
+        $this->places[$place] -= $nbToken;
+
+        if (0 > $this->places[$place]) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" could not contain a negative token number.', $place));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/Marking.php]: I suppose this check should happen before the subtraction so that we don't mutate the object and then throw

Comment (lyrixx) [reply src/Symfony/Component/Workflow/Marking.php]: Very good catch, thanks. Fixed

File: src/Symfony/Component/Workflow/Marking.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Workflow/Marking.php]: consistency > personal taste ;)

File: src/Symfony/Component/Workflow/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Add method `getEnabledTransition()` to `WorkflowInterface`
  * Automatically register places from transitions
+ * Add support for workflow that need to store many tokens in the marking
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/CHANGELOG.md]:
```suggestion
 * Add support for workflows that need to store many tokens in the marking
```

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Workflow/CHANGELOG.md (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/CHANGELOG.md]: Oups, thanks, fixed

File: src/Symfony/Component/Workflow/Marking.php [hunk 4]
```diff
@@ -27,23 +27,58 @@ class Marking
...
     {
         foreach ($representation as $place => $nbToken) {
-            $this->mark($place);
+            $this->mark($place, $nbToken);
         }
     }
 
-    public function mark(string $place): void
+    /**
+     * @param int $nbToken
```
Comment (stof) [inline src/Symfony/Component/Workflow/Marking.php:35]: I suggest adding `@psalm-param int<1, max> $nbToken` (or `@psalm-param positive-int $nbToken` which is an alias for that range) to help static analyzers detect invalid usages of the argument (this is added in `@psalm-param` for now as we need `@param` to specify the future native type as long as we don't have a proper parameter for it yet).

File: src/Symfony/Component/Workflow/Marking.php [hunk 5]
```diff
@@ -27,23 +27,58 @@ class Marking
...
+     */
+    public function mark(string $place /* , int $nbToken = 1 */): void
     {
-        $this->places[$place] = 1;
+        $nbToken = 1 < \func_num_args() ? func_get_arg(1) : 1;
+
+        if ($nbToken < 1) {
+            throw new \InvalidArgumentException(sprintf('The number of tokens must be greater than 0, "%s" given.', $nbToken));
+        }
+        dump(isset($this->place[$place]));
```
Comment (stof) [inline src/Symfony/Component/Workflow/Marking.php]: `dump()` must be removed

File: src/Symfony/Component/Workflow/Marking.php [hunk 6]
```diff
@@ -27,23 +27,58 @@ class Marking
...
+            unset($this->places[$place]);
+        } else {
+            $this->places[$place] = $nbToken;
+        }
     }
 
     public function has(string $place): bool
     {
-        return isset($this->places[$place]);
+        return \array_key_exists($place, $this->places);
```
Comment (stof) [inline src/Symfony/Component/Workflow/Marking.php]: why changing that ?

File: src/Symfony/Component/Workflow/Marking.php [hunk 7]
```diff
@@ -27,23 +27,58 @@ class Marking
...
+        }
+
+        if (!$this->has($place)) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" is not marked.', $place));
+        }
+
+        $nbToken = $this->places[$place] - $nbToken;
+
+        if (0 > $nbToken) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" could not contain a negative token number.', $place));
```
Comment (stof) [inline src/Symfony/Component/Workflow/Marking.php]: I suggest rewording the message to say that you cannot unmark the place from more token than its marking, perhaps mentioning the existing marking number and the number of removed token. It might make the error easier to understand.

File: src/Symfony/Component/Workflow/Marking.php [hunk 8]
```diff
@@ -27,23 +27,58 @@ class Marking
...
+
+        if ($nbToken < 1) {
+            throw new \InvalidArgumentException(sprintf('The number of tokens must be greater than 0, "%s" given.', $nbToken));
+        }
+
+        if (!$this->has($place)) {
+            throw new \InvalidArgumentException(sprintf('The place "%s" is not marked.', $place));
+        }
+
+        $nbToken = $this->places[$place] - $nbToken;
```
Comment (stof) [inline src/Symfony/Component/Workflow/Marking.php]: I would suggest to use separate variable to represent the number of removed tokens and the number of tokens left in the place. It would make the code easier to understand than re-purposing the same variable (and will be needed anyway if you want to include the number of removed tokens in the error message)

File: src/Symfony/Component/Workflow/Tests/MarkingTest.php [hunk 9]
```diff
@@ -22,24 +22,70 @@ public function testMarking()
...
+    public function testGuardNotNbTokenLowerThanZero()
+    {
+        $marking = new Marking(['a' => 1]);
+
+        $this->expectException(\LogicException::class);
+        $this->expectExceptionMessage('The place "a" could not contain a negative token number.');
+        $marking->unmark('a', 2);
+    }
+
+    public function testGuardNotNbTokenEquals0()
```
Comment (stof) [inline src/Symfony/Component/Workflow/Tests/MarkingTest.php]: I suggest renaming those tests to make their intent clearer:

- they don't mention that they cover the `unmark` test (and we might need a similar test cover `->mark('a', 0)`)
- `GuardNotNbTokenLowerThanZero` could be easier to understand if you describe it as unmarking more token than the current state.

File: src/Symfony/Component/Workflow/Marking.php (hunk 6 above)
Comment (lyrixx) [reply src/Symfony/Component/Workflow/Marking.php]: reverted

**Comment by lyrixx** [comment]
@stof Thanks for the review, I addressed your comments

**Review by stof — APPROVED** [review]

**Comment by stof** [comment]
@lyrixx please fix the coding standard issues reported by fabbot (for the exception formatting ones, I think they can be considered false positives)

**Comment by lyrixx** [comment]
@stof, thanks, fixed
